src/binary-encoding.adoc

```diff
@@ -0,0 +1,1618 @@
+= Binary Encoding
+:toc:
+
+[[encoding_primitives]]
+== Encoding primitives
+
+[[varuint]]
+=== `VarUInt`
+
+A variable-length unsigned integer.
+
+The bytes of a ``VarUInt``s are written in
+link:https://en.wikipedia.org/wiki/Endianness:[Little Endian byte order]. This means that the first byte will contain
+the ``VarUInt``'s least significant bits.
+
+The least significant bits in the `VarUInt` indicate the number of bytes that were used to encode the integer.
+If a `VarUInt` is `_N_` bytes long, its `_N-1_` least significant bits will be `0`; a terminal `1` bit will be
+in the next most significant position.
+All bits that are more significant than the terminal `1` represent the magnitude of the `VarUInt`.
+
+==== Example encoding of VarUInt 14 ====
+[source]
+----
+              ┌──── Lowest bit is 1 (`end`), indicating
+              │     this is the only byte.
+0 0 0 1 1 1 0 1
+└─────┬─────┘
+unsigned int 14
+----
+
+==== Example encoding of VarUInt 729 ====
+[source]
+----
+             ┌──── There's 1 zero in the least significant bits, so this
+             │     integer is two bytes wide.
+            ┌┴┐
+0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
+└────┬────┘      └──────┬──────┘
+lowest 6 bits    highest 8 bits
+of the unsigned  of the unsigned
+integer          integer
+----
+
+==== Example encoding of VarUInt 21,043 ====
+[source]
+----
+            ┌───── There are 2 zeros in the least significant bits, so this
+            │      integer is three bytes wide.
+          ┌─┴─┐
+1 0 0 1 1 1 0 0  1 0 0 1 0 0 0 1  0 0 0 0 0 0 1 0
+└───┬───┘        └──────┬──────┘  └──────┬──────┘
+lowest 6 bits    next 8 bits of   highest 8 bits
+of the unsigned  the unsigned     of the unsigned
+integer          integer          integer
+----
+
+NOTE: Ion 1.0 uses link:https://en.wikipedia.org/wiki/Endianness[Big Endian byte order] and indicates the width of the
+integer by having
+link:https://amazon-ion.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields[an `end` flag bit in the highest
+ bit of each byte].
```

Comment:
Initial versions of this doc included lots of comparisons with the 1.0 encodings, but I found it to be pretty distracting from the core content. I think I'll remove this note, especially since we're going to rename `VarUInt`/`VarInt` in this version.